llvm/lib/Target/AArch64/AArch64SVEInstrInfo.td

```diff
@@ -2583,6 +2583,27 @@ let Predicates = [HasSVEorSME] in {
                                sub_32)>;
   }
 
+  // Add NoUseScalarIncVL to avoid affecting for patterns with UseScalarIncVL
+  let Predicates = [NoUseScalarIncVL] in {
+    def : Pat<(add GPR64:$op, (vscale (sve_rdvl_imm i32:$imm))),
+              (ADDXrs GPR64:$op, (RDVLI_XI $imm), 0)>;
```

Comment:
Yes, it is need. Without this, the case https://gcc.godbolt.org/z/TndasfP6h will transform into the following code because we matched the new added pattern `def : Pat<(add GPR64:$op, (vscale (sve_cnth_imm_neg i32:$imm)))`
```
cnth   x8, all, mul #2
sub     x0, x0, w8
```

Reply:
I agree there's a code difference but it's not bad is it? They're both result in two instructions, with the first being invariant and thus hoist able.

Reply:
Yes, delete the pattern `(vscale (sve_rdvl_imm i32:$imm)`, thanks
I think `cnth   x8, all, mul #2` will need an extra `mull unit` before, but it indeed doesn't after checking with with colleagues